### PR TITLE
#51: fix terminal multiplexers bypass remote sudo password

### LINKED ISSUE #51: terminal multiplexers bypass remote sudo password
hi me again, thanks for recent update, i just happen to be running version 0.7.1 and version 0.7.3 at the same time, so I can confirm this bug also exists on version 0.7.1
basically the steps to reproduce it are these:
- ssh into a remote system
- run tmux/byobu
- run sudo su

deny remote is set to true and it indeed denies access outside of byobu, it also denies the 'cup of tee' access aswell in the case of 0.7.3(as expected), but inside of byobu or tmux it happily grants access

### PR DESCRIPTION
**Pull request by mcdope**
The tty only approach had downsides (see #8), its fix had downsides too (see #39), and the name approach had downsides (see #51) too, let's try all together plus some additional magic.

This modifies local check to:
* check for sshd/telnetd in process chain - for the obvious cases
* check for tmux in process chain, if found parse its environment to determine tmux client id to determine session tty for utmp check
* if tmux found, but session tty not, check for remote clients attached to tmux
* check for DISPLAY, if found use that for utmp check
* in case no remote daemon was found, tmux wasn't detected, and DISPLAY is not set - fall back to good ol' ttyname() which should now be safe since we handled all edge cases before

Closes #51

### CODE REVIEW

File: src/local.c [hunk 1]
```diff
@@ -31,20 +33,121 @@ int pusb_local_login(t_pusb_options *opts, const char *user)
...
+	endutent();
+
+	if (!utent)
+	{
+		log_error("No utmp entry found for tty \"%s\", assuming remote\n", utsearch.ut_line);
+		return (0);
+	} else {
+		log_debug("		utmp entry found\n");
+		log_debug("			utmp->ut_pid: %d\n", utent->ut_pid);
+		log_debug("			utmp->ut_user: %s\n", utent->ut_user);
```
Comment (MarvinFollmann) [inline src/local.c:131]: nit: maybe think about having an optional parameter for indentation

File: src/local.c [hunk 2]
```diff
@@ -31,20 +34,144 @@ int pusb_local_login(t_pusb_options *opts, const char *user)
...
+		if (!has_xdmcp) {
+			/** 
+			 * If we reach this point there is neither DISPLAY set so we can't assume a tty to be set
+			 * It will most likely be a graphical login session, since we checked if XDMCP is enabled 
+			 * before it should be safe to allow the request. 
+			 **/
+			log_debug("		Loginmanager but no DISPLAY env var detected, assuming graphical login session.\n");
+			return (1);
+		}
+	} else if (display == NULL) {
```
Comment (mcdope) [inline src/local.c:127]: make just else

File: src/local.c [hunk 3]
```diff
@@ -31,20 +34,144 @@ int pusb_local_login(t_pusb_options *opts, const char *user)
...
+		strncpy(utsearch.ut_line, tmux_client_tty, sizeof(utsearch.ut_line) - 1);
+	} else if (display != NULL) {
+		// No tmux, Xsession detected, set DISPLAY in utsearch
+		log_debug("		Using DISPLAY %s for search\n", display);
+		strncpy(utsearch.ut_line, display, sizeof(utsearch.ut_line) - 1);
+	} else if (has_loginmanager) {
+		// Call from loginmanager detected, previous methods failed
+		if (!has_xdmcp) {
+			/** 
+			 * If we reach this point there is neither DISPLAY set so we can't assume a tty to be set
```
Comment (mcdope) [inline src/local.c:120]: fix comment

File: src/port_check.c [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#include <sys/socket.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <arpa/inet.h>
+#include "log.h"
+
+#define XDMCP_PORT 177
+
+static const int UDP_CONNECT_TIMEOUT = 250 * 100000;
```
Comment (mcdope) [inline src/port_check.c:33]: 🤦 fix

File: src/port_check.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    if ( (sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) {
+        log_debug("UDP socket socket creation failed!\n");
+        return 0;
+    }
+
+    // Set timeout
+    struct timeval tv;
+    tv.tv_sec = 0;
+    tv.tv_usec = UDP_CONNECT_TIMEOUT;
+    if (setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) < 0) {
```
Comment (mcdope) [inline src/port_check.c:58]: failed in vm - why

File: src/port_check.c [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        sockfd,
+        (const char *)content,
+        strlen(content),
+        MSG_CONFIRM,
+        (const struct sockaddr *) &servaddr,
+        sizeof(servaddr)
+    );
+
+    // Wait till we got a response (or timeout kicks in)
+    n = recvfrom(
```
Comment (mcdope) [inline src/port_check.c:82]: MSG_WAITALL really the right one?

File: src/local.c [hunk 7]
```diff
@@ -18,33 +18,249 @@
...
+	struct dirent *dent_proc;
+	while ((dent_proc = readdir(d_proc)) != NULL)
+	{
+		if (dent_proc->d_type == DT_DIR && atoi(dent_proc->d_name) != 0 && strcmp(dent_proc->d_name, ".") != 0 && strcmp(dent_proc->d_name, "..") != 0)
+		{
+			memset(cmdline_path, 0, 32);
+			sprintf(cmdline_path, "/proc/%s/cmdline", dent_proc->d_name);
+
+			memset(cmdline, 0, 4096);
+			int cmdline_file = open(cmdline_path, O_RDONLY | O_CLOEXEC);
```
Comment (mcdope) [inline src/local.c:98]: refactor to fopen/fread/fclose to keep it uniform

File: src/local.c [hunk 8]
```diff
@@ -18,33 +18,249 @@
...
+		get_process_parent_id(pid, & pid);
 		if (strstr(name, "sshd") != NULL || strstr(name, "telnetd") != NULL) {
 			log_error("One of the parent processes found to be a remote access daemon, denying.\n");
 			return (0);
 		}
 	}
 
-	log_debug("No remote daemons found in parent process list, seems to be local request - allowing.\n");
-	return (1);
+	if (strcmp(service, "gdm-password") == 0 || strcmp(service, "xdm") == 0 || strcmp(service, "lightdm") == 0) {
```
Comment (mcdope) [inline src/local.c:202]: test whitelisting for sddm, will need an additional condition most likely

File: src/process.c [hunk 9]
```diff
@@ -73,4 +73,41 @@ void get_process_parent_id(const pid_t pid, pid_t * ppid) {
...
+
+/**
+ * Read environment variable of another process
+ *
+ * @param pid pid of process to read the environment of
+ * @param var envvar to look up
+ *
+ * @return content of var if found, else NULL
+ */
+char *get_process_envvar(pid_t pid, char *var)
```
Comment (mcdope) [inline src/process.c:86]: add pusb_ prefix

File: src/process.h [hunk 10]
```diff
@@ -29,4 +29,6 @@ void get_process_name(const pid_t pid, char * name);
 
 void get_process_parent_id(const pid_t pid, pid_t * ppid);
```
Comment (mcdope) [inline src/process.h]: add pusb_ prefix

File: src/process.h [hunk 11]
```diff
@@ -29,4 +29,6 @@ void get_process_name(const pid_t pid, char * name);
...
 
+char *get_process_envvar(pid_t pid, char *var);
```
Comment (mcdope) [inline src/process.h]: add pusb_ prefix

File: src/tmux.c [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+ * Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "log.h"
+#include "process.h"
+
+char *tmux_get_client_tty(pid_t env_pid)
```
Comment (mcdope) [inline src/tmux.c]: add pusb_ prefix

File: src/tmux.h [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
+ * Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef PUSB_TMUX_H_
+# define PUSB_TMUX_H_
+
+char *tmux_get_client_tty(pid_t tmux_pid);
```
Comment (mcdope) [inline src/tmux.h]: add pusb_ prefix

File: debian/changelog [hunk 14]
```diff
@@ -1,3 +1,9 @@
+libpam-usb (0.7.4) unstable; urgency=high
+
+  * Massive rework of local check / deny_remote
```
Comment (mcdope) [inline debian/changelog:3]: needs more details, also has other fixes missing. best to remove and bump version in a separate PR

File: debian/libpam-usb.NEWS [hunk 15]
```diff
@@ -1,4 +1,16 @@
-libpam-usb (0.7.0) focal; urgency=medium
+libpam-usb (0.7.4) unstable; urgency=high
+
+  Reworked deny_remote option / local check. It now tries all known
```
Comment (mcdope) [inline debian/libpam-usb.NEWS:3]: see above

File: src/local.c [hunk 16]
```diff
@@ -18,33 +18,258 @@
...
+		pusb_get_process_parent_id(pid, & pid);
 		if (strstr(name, "sshd") != NULL || strstr(name, "telnetd") != NULL) {
 			log_error("One of the parent processes found to be a remote access daemon, denying.\n");
 			return (0);
 		}
 	}
 
-	log_debug("No remote daemons found in parent process list, seems to be local request - allowing.\n");
-	return (1);
+	if (strcmp(service, "gdm-password") == 0 || strcmp(service, "xdm") == 0 || strcmp(service, "lightdm") == 0 || strcmp(service, "sddm") == 0) {
```
Comment (mcdope) [inline src/local.c:202]: add a warning to documentation to not combine pamusb witth remote x login

Comment (mcdope) [reply src/local.c:202]: https://github.com/mcdope/pam_usb/wiki/Security

File: src/local.c (hunk 7 above)
Comment (mcdope) [reply src/local.c:98]: will stay like this for now. this is tested, and the `\0`s in the file being read could maybe cause issues.

**Pull request merged**